### DISCUSSION TITLE
v4.0.0

### DISCUSSION THREAD

**Discussion by joschi** [dropwizard/dropwizard#6931] [Announcements]
# Dropwizard 4.0.0

It's finally here, the next major release of Dropwizard! This one has been developed and released in tandem with [Dropwizard 3.0.0](https://github.com/dropwizard/dropwizard/releases/tag/v3.0.0).

If you're wondering what the differences are: Not much!

3️⃣ Dropwizard 3.0.0 is still based on and is using dependencies for Java EE and the `javax.*` package namespace. The effort to migrate from Dropwizard 2.x to Dropwizard 3.0.0 should be minimal for many projects.

4️⃣ Dropwizard 4.0.0 is based on and is using dependencies for Jakarta EE and the `jakarta.*` package namespace. The effort to migrate from Dropwizard 2.x to Dropwizard 4.0.0 might be bigger due to more package changes and more breaking changes in our dependencies.

📣 In any case, let us know how the migration went and if you were missing anything important in the upgrade notes in our [GitHub Discussions](https://github.com/dropwizard/dropwizard/discussions)!

## ⚠️  Breaking changes

For a full list of relevant changes please refer to the [Upgrade Notes for Dropwizard 4.0.x](https://www.dropwizard.io/en/release-4.0.x/manual/upgrade-notes/upgrade-notes-4_0_x.html).

### Java 11 baseline

Dropwizard 4.0.0 and later require Java 11 as a baseline. Later versions of Java are also working.

### Dropwizard Package Structure and JPMS

In order to properly support the Java Platform Module System (JPMS), the Java packages in modules must not overlap, or put differently, the packages may not be split into multiple modules.

Dropwizard 4.0.0 won’t enable full support for the JPMS. Instead, as a transition step, automatic modules are introduced.

Affected packages:

Maven module | Old package | New package
-- | -- | --
`dropwizard-core` | `io.dropwizard` | `io.drowizard.core`
`dropwizard-logging` | `io.dropwizard.logging` | `io.dropwizard.logging.common`
`dropwizard-metrics` | `io.dropwizard.metrics` | `io.dropwizard.metrics.common`
`dropwizard-views` | `io.dropwizard.views` | `io.dropwizard.views.common`

### Transition to Jakarta EE

Previously released Dropwizard versions used Java/Jakarta EE dependencies under the `javax` namespace. Dropwizard 4.0.x transitioned to Jakarta EE 9 components and therefore utilize the new `jakarta` namespace for many components.

This means that most of the imports of existing applications using Dropwizard 2.x or 3.x will have to be changed from the `javax` to the `jakarta` namespace. However, other components still use the `javax` namespace, so a simple search and replace could break other imports.

### Jakarta EE 9 compatibility

As stated above, Dropwizard 4.0.x will transition to Jakarta EE 9 components. This means, Dropwizard will now try to be consistent with one specific EE version. 

Therefore Dropwizard 4.0.x will stay on components of Jakarta EE 9 and the transition to components of Jakarta EE 10 will be postponed to an other release series. 

Since the EE version bump will probably introduce breaking changes, the Jakarta EE 10 components will be most likely integrated in Dropwizard 5.0.x and not in a 4.1.x release.

### Jetty 11

The main change introduced in Dropwizard 4.0.0 is the upgrade to Jetty 11.0.x. Jetty 11.0.x is built for Java 11, therefore the Java version change was necessary.

This change comes with some migration cost. For detailed information regarding the changes introduced in Jetty 11.0.x, refer to the [Jetty migration guide](https://www.eclipse.org/jetty/documentation/jetty-10/programming-guide/index.html#pg-migration-94-to-11).

### Apache HttpClient 5

The version of Apache HttpClient used in [dropwizard-client](https://www.dropwizard.io/en/release-4.0.x/manual/client.html) has been updated to version 5.x.

For more information refer to the [Apache HttpClient 5.0 migration guide](https://hc.apache.org/httpcomponents-client-5.1.x/migration-guide/migration-to-classic.html).

### Hibernate 6

Hibernate 5.6 provides compatible implementations for JPA 2.2 and for Jakarta Persistence 3.0.

But Hibernate 6.0 and 6.1 still provide compatible implementations for Jakarta Persistence 3.0, so we upgraded to Hibernate 6.1 in Dropwizard 4.0.x.

If any other aspects from Hibernate are used in an application rather than those provided by Dropwizard, there might be an additional migration cost. Please follow the [Hibernate 6 migration guide](https://github.com/hibernate/hibernate-orm/blob/6.0/migration-guide.adoc) and the [Hibernate 6.1 migration guide](https://github.com/hibernate/hibernate-orm/blob/6.1/migration-guide.adoc), if you encounter any problems.

### Removed Support for JUnit 4.x

Support for testing with JUnit 4.x has been moved from `dropwizard-testing` to [dropwizard-testing-junit4](https://github.com/dropwizard/dropwizard-testing-junit4).

## 🥇 Special thanks

The release of Dropwizard 3.0.0 wouldn't have been possible without the help of the community (YOU!) and especially the following contributors:

@zUniQueX, @rhowe, @pstackle, @dennyac, and countless others!

## What's Changed since Dropwizard 4.0.0-beta.4
* Update dependency org.apache.tomcat:tomcat-jdbc to v10.1.6 by @renovate in https://github.com/dropwizard/dropwizard/pull/6743
* Update github/codeql-action digest to 32dc499 by @renovate in https://github.com/dropwizard/dropwizard/pull/6742
* Update dependency org.eclipse.jetty:jetty-bom to v11.0.14 by @renovate in https://github.com/dropwizard/dropwizard/pull/6755
* Update dependency org.apache.maven.plugins:maven-compiler-plugin to v3.11.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6756
* Update dependency org.apache.maven.plugins:maven-compiler-plugin to v3.11.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6757
* Update scacap/action-surefire-report digest to a2d14ba by @renovate in https://github.com/dropwizard/dropwizard/pull/6754
* Update dependency org.checkerframework:checker-qual to v3.32.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6766
* Properly declare plugin versions in java-simple archetype by @joschi in https://github.com/dropwizard/dropwizard/pull/6769
* Update dependency org.apache.tomcat:tomcat-jdbc to v10.1.7 by @renovate in https://github.com/dropwizard/dropwizard/pull/6771
* Update dependency com.github.ben-manes.caffeine:caffeine to v3.1.5 by @renovate in https://github.com/dropwizard/dropwizard/pull/6773
* Update dependency net.bytebuddy:byte-buddy to v1.14.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6778
* Update dependency com.uber.nullaway:nullaway to v0.10.10 by @renovate in https://github.com/dropwizard/dropwizard/pull/6782
* Update actions/cache action to v3.3.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6792
* Update mockito monorepo to v5.2.0 (minor) by @renovate in https://github.com/dropwizard/dropwizard/pull/6793
* Update github/codeql-action digest to 16964e9 by @renovate in https://github.com/dropwizard/dropwizard/pull/6804
* Update actions/cache action to v3.3.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6805
* Update scacap/action-surefire-report digest to 4845216 by @renovate in https://github.com/dropwizard/dropwizard/pull/6817
* Update dependency maven to v3.8.8 by @renovate in https://github.com/dropwizard/dropwizard/pull/6818
* Update dependency net.bytebuddy:byte-buddy to v1.14.2 by @renovate in https://github.com/dropwizard/dropwizard/pull/6819
* Update dependency maven-wrapper to v3.2.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6820
* Update dependency org.apache.maven.plugins:maven-surefire-plugin to v3 by @renovate in https://github.com/dropwizard/dropwizard/pull/6821
* Update dependency org.apache.maven.plugins:maven-surefire-plugin to v3 by @renovate in https://github.com/dropwizard/dropwizard/pull/6822
* Update logback.version to v1.4.6 (patch) by @renovate in https://github.com/dropwizard/dropwizard/pull/6836
* Update actions/checkout digest to 24cb908 by @renovate in https://github.com/dropwizard/dropwizard/pull/6834
* Update github/codeql-action digest to 168b99b by @renovate in https://github.com/dropwizard/dropwizard/pull/6835
* Update actions/checkout action to v3.4.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6837
* Update slf4j.version to v2.0.7 (patch) by @renovate in https://github.com/dropwizard/dropwizard/pull/6839
* Update dependency io.dropwizard.metrics:metrics-bom to v4.2.18 by @renovate in https://github.com/dropwizard/dropwizard/pull/6847
* Update dependency maven to v3.9.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6848
* Update dependency org.apache.maven.plugins:maven-help-plugin to v3.4.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6849
* Update dependency com.mattbertolini:liquibase-slf4j to v5 by @renovate in https://github.com/dropwizard/dropwizard/pull/6854
* Update dependency org.apache.maven.plugins:maven-release-plugin to v3 by @renovate in https://github.com/dropwizard/dropwizard/pull/6855
* Update github/codeql-action digest to 67a35a0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6859
* Update actions/stale action to v8 by @renovate in https://github.com/dropwizard/dropwizard/pull/6864
* Use Mockito BOM in dropwizard-dependencies by @joschi in https://github.com/dropwizard/dropwizard/pull/6867
* Update actions/checkout digest to 8f4b7f8 by @renovate in https://github.com/dropwizard/dropwizard/pull/6894
* Update dependency org.apache.maven.plugins:maven-deploy-plugin to v3.1.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6895
* Update dependency org.apache.maven.plugins:maven-deploy-plugin to v3.1.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6896
* Update dependency org.apache.maven.plugins:maven-install-plugin to v3.1.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6897
* Update dependency org.apache.maven.plugins:maven-install-plugin to v3.1.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6898
* Update dependency org.apache.maven.plugins:maven-resources-plugin to v3.3.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6899
* Update dependency org.apache.maven.plugins:maven-resources-plugin to v3.3.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6900
* Update dependency org.apache.maven.shared:maven-filtering to v3.3.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6901
* Update actions/checkout action to v3.5.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6902
* Set application state healthy if there are no unhealthy critical checks by @joschi in https://github.com/dropwizard/dropwizard/pull/6908
* Update actions/setup-java digest to 5ffc13f by @renovate in https://github.com/dropwizard/dropwizard/pull/6914
* Update github/codeql-action digest to 04df126 by @renovate in https://github.com/dropwizard/dropwizard/pull/6915
* Update dependency net.bytebuddy:byte-buddy to v1.14.3 by @renovate in https://github.com/dropwizard/dropwizard/pull/6919

**Full Changelog**: https://github.com/dropwizard/dropwizard/compare/v4.0.0-beta.4...v4.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/dropwizard/dropwizard/releases/tag/v4.0.0'>v4.0.0</a>.</em>

**Comment by cw-dimedis** [comment]
I just upgraded 2 projects from 2.0.x and 2.1.x to 4.0.0. The migration was a breeze, no problems at all.

Note: If you are using the dependency "swagger-jaxrs2", use the artifact "swagger-jaxrs2-jakarta" instead.

**Comment by cw-dimedis** [comment]
I just noticed this warning when starting the service:
`WARN org.hibernate.validator.internal.properties.javabean.JavaBeanExecutable: HV000254: Missing parameter metadata for ResponseMeteredLevel(String, int), which declares implicit or synthetic parameters. Automatic resolution of generic type information for method parameters may yield incorrect results if multiple parameters have the same erasure. To solve this, compile your code with the '-parameters' flag.`

Might be a bug in Hibernate Validator: https://hibernate.atlassian.net/browse/HV-1852

Setting the "parameters" option in the Maven Compiler Plugin as suggested on StackOverflow doesn't fix it.

Do you think we can safely ignore this warning?

**Reply by cw-dimedis** [reply]
Just checked: I get the same warning when I spin up the HelloWorldApplication from dropwizard-example (using OpenJDK Runtime Environment Temurin-11.0.17+8)

**Comment by natan-abolafya** [comment]
Hello. I'm a bit confused if dropwizard 4.0 has jetty 10 or 11.

It looks like 11 but this part got copied from 3.0 release and not updated for 4.0. Am I assuming it right?
![image](https://github.com/dropwizard/dropwizard/assets/64091305/1b1d6087-3145-4364-9a42-03aebbeb49d5)

**Reply by joschi** [reply]
@natan-abolafya Yes, you are absolutely correct!

This being said, the only difference between Jetty 10 and 11 is the usage of the Jakarta EE namespace.

See also https://webtide.com/jetty-10-and-11-have-arrived/

> ## Jetty 11
> When Oracle donated the JavaEE project to the Eclipse foundation, they maintained the copyright to the `javax.*` namespace. Subsequently, future versions of JavaEE, now called JakartaEE, would be required to use a new namespace for those packages. Instead of an incremental process, the Eclipse foundation opted for a big bang approach and packages formerly under the `javax.*` namespace has been refactored under the new `jakarta.*` namespace.
> 
> Enter Jetty 11. _**Jetty 11 is identical to Jetty 10 except that the `javax.*` packages now conform to the new `jakarta.*` namespace.**_ Jetty 10 and 11 will remain in lockstep with each other for releases, meaning all new features or bug fixes in one version will be available in the other.
> 
> It is important to note that Jetty 10.x will be the last major Jetty version to support the `javax.*` namespace. If this is alarming, be assured that Jetty 10 will be supported for a number of years to come. We have no plans on releasing it only to drop support for it in 12-18 months.

**Reply by natan-abolafya** [reply]
Great, thanks! Really appreciate the move with 3.0 and 4.0 by the way. Makes the upgrade step by step easier.

**Comment by btlorch** [comment]
Thanks for your continuous efforts and latest improvements to Dropwizard. While trying to upgrade from an older version, I ran into a problem with logging. My *build.gradle* contains dropwizard-logging 4.0.0 as a dependency. The problem is that the server immediately terminates with the following exception:
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
```

The referenced slfj4 docs state that placing one of slf4j-nop.jar, slf4j-simple.jar, slf4j-log4j12.jar, slf4j-jdk14.jar or logback-classic.jar should solve the problem. I tried adding several of them (one at a time) as a dependency to my *build.gradle*, but keep getting the same exception. Is there a simple solution, without resorting to external logging?

**Reply by zUniQueX** [reply]
Hi @btlorch. The `StaticLoggerBinder` is part of SLF4J 1.7. Dropwizard 4 uses SLF4J 2.x, which relies on the `ServiceLoader` mechanism to load an implementation. Please make sure to use SLF4J 2.x in your project.

Dropwizard internally uses Logback as the default SLF4J implementation. So I'd suggest plaching logback-classic (version 1.4.x) on your classpath.

**Reply by btlorch** [reply]
Thanks for the explanation! It turned out that one of my dependencies was a JAR file that included SLF4J 1.7. Recompiling the JAR file with SLF4J excluded solve it for me.

**Comment by cw-dimedis** [comment]
Hi dropwizard dudes, ;)

we just noticed that the dropwizard-gelf AppenderFactory (https://github.com/gini/dropwizard-gelf) is not loaded anymore in DW4, so we can't configure logging against Graylog. Is that to be expected? And what would be ways around that?

Any input is appreciated. Thanks,
Christoph

**Reply by joschi** [reply]
The last version of dropwizard-gelf was released for Dropwizard 2.0.x which itself is EOL since January 31, 2023.

Since this project mostly a wrapper around https://github.com/mp911de/logstash-gelf, you could use the Logback appender of that project directly.

Other than that, using https://github.com/logfellow/logstash-logback-encoder might also be a viable alternative to get logs into Graylog (Raw TCP input and parse and flatten the JSON payload).

**Reply by cw-dimedis** [reply]
Thanks for the reply.

It seems like many of the Java projects supporting GELF have gone out of fashion. logstash-gelf is also unsupported as of last month and the last release has a few CVEs in its the dependencies.

Thanks for the hint regarding logstash-logback-encoder, I will look into that.

But maybe it's not really best practice anymore to log directly to Graylog from the Java service? I will also look into using a collector sidecar instead.

Cheers
Christoph

**Comment by Sachin-Bhararti** [comment]
Hello dropwizard team!

We are currently in the process of moving from Dropwizard 2 -> Dropwizard 4. We noticed that our `@Annotations` no longer seem to be working after the move. An example of this is having a method in dropwizard 2. We are a java17 application and use gradle as our build tool.


```
import javax.annotation.Nonnull;
public void foo(@Nonnull int x) {

}
```
When upgraded to Dropwizard 4 no longer seems to throw an exception when the function is called with a null parameter

```
import jakarta.annotation.Nonnull;
public void foo(@Nonnull int x) {

}
```

Trying to debug this issue has been tough as reading the documentation it seems that it would be expected that this would work out of the box https://www.dropwizard.io/en/stable/manual/validation.html

We have tried to use the @NotNull annotation that belongs to the jakarta.validation library but that too seems to not be plumbed together.

An observation we had is that Dropwizard 4.0 now uses hibernate-validator version 8.0.1.Final but the release notes say that Dropwizard is compatible with hibernate 6. https://www.dropwizard.io/en/stable/manual/upgrade-notes/upgrade-notes-4_0_x.html

It looks like the Environment.getValidator() returns a jakarta.Validator and that the hibernate-validator 8 also accepts this so it doesnt look like there is an issues here.

Another observation is that [dropwizard 4](https://www.dropwizard.io/en/stable/manual/upgrade-notes/upgrade-notes-4_0_x.html) looks like its using JakartaEE 9 and [hibernate 8](https://hibernate.org/validator/releases/8.0/) seems to support Jakarta 10. [Hibernate 7](https://hibernate.org/validator/releases/7.0/) uses JakartaEE 9 but overriding the build file to use it doesnt solve the issue

Has anyone had any issues similar to the one we are facing? And if they have is there a known issue?


If not is there configuration code I can be pointed towards where these annotations would be configured at run time so I can take a deeper look there into what may be happening.

Thanks!

**Reply by zUniQueX** [reply]
Hi @Sachin-Bhararti. To validate objects, you should use the `jakarta.validation` annotations. As you correctly pointed out, these ones are evaluated by a `Validator` which then eventually raises an exception, if the constraint is violated.

The release notes only refer to the term 'Hibernate'. This is commonly used for the Hibernate ORM database library. Hibernate Validator will always be refered to with the term 'Hibernate Validator'. That being said, Dropwizard 4.0.x is compatible with Hibernate ORM 6.1 and _not_ Hibernate Validator 6.1.

The artifact of Hibernate Validator is currently shipped with the version `7.0.5.Final`, which is the EE9 compatible version. Hibernate Validator 8 is EE10 only and therefore not used in Dropwizard 4.0.x. The Hibernate Validator 8 artifact in your application is probably pulled in from an other dependency. Note that this artifact requires a separate EL implementation to be present on the classpath at runtime.

**Reply by Sachin-Bhararti** [reply]
Updating to 7.0.5 resolved this issue for us
